optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/BiAnchor.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.bi;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.ToIntBiFunction;
+import java.util.function.ToLongBiFunction;
+
+import org.drools.model.Declaration;
+import org.drools.model.Drools;
+import org.drools.model.Global;
+import org.drools.model.PatternDSL;
+import org.drools.model.RuleItemBuilder;
+import org.drools.model.consequences.ConsequenceBuilder;
+import org.drools.model.functions.Block4;
+import org.kie.api.runtime.rule.RuleContext;
+import org.optaplanner.core.api.score.holder.AbstractScoreHolder;
+import org.optaplanner.core.impl.score.stream.common.JoinerType;
+import org.optaplanner.core.impl.score.stream.drools.common.InferredRuleMetadata;
+import org.optaplanner.core.impl.score.stream.drools.common.LogicalTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.GenuineRuleMetadata;
+import org.optaplanner.core.impl.score.stream.drools.common.RuleMetadata;
+import org.optaplanner.core.impl.score.stream.drools.tri.TriAnchor;
+import org.optaplanner.core.impl.score.stream.drools.uni.UniAnchor;
+import org.optaplanner.core.impl.score.stream.tri.AbstractTriJoiner;
+
+import static org.drools.model.DSL.on;
+
+public final class BiAnchor {
```

Comment:
Knee jerk reaction: Anchor is a name already used in the context of chained variables. Can we use anything else?